src/System.ValueTuple/src/System/ValueTuple/ValueTuple.cs

```diff
@@ -2060,12 +2269,43 @@ string ITupleInternal.ToStringEnd()
             }
         }
 
-        int ITupleInternal.Size
+        /// <summary>
+        /// The number of positions in this data structure.
+        /// </summary>
+        int ITuple.Length
+        {
+            get
+            {
+                return 7 + ((ITuple)Rest).Length;
+            }
+        }
+
+        /// <summary>
+        /// Get the element at position <param name="index"/>.
+        /// </summary>
+        object ITuple.this[int index]
         {
             get
             {
-                ITupleInternal rest = Rest as ITupleInternal;
-                return rest == null ? 8 : 7 + rest.Size;
+                switch (index)
+                {
+                    case 0:
+                        return Item1;
+                    case 1:
+                        return Item2;
+                    case 2:
+                        return Item3;
+                    case 3:
+                        return Item4;
+                    case 4:
+                        return Item5;
+                    case 5:
+                        return Item6;
+                    case 6:
+                        return Item7;
+                }
+
+                return ((ITuple)Rest)[index - 7];
```

Comment:
Casting to `ITuple` is going to box `Rest`. The box allocation can be avoided by doing:

```c#
return Rest[index - 7];
```